tidy3d/plugins/adjoint/components/data/data_array.py

```diff
@@ -312,28 +313,37 @@ def isel(self, **isel_kwargs) -> JaxDataArray:
 
         return self_sel
 
-    def sel(self, indexers: dict = None, method: str = "nearest", **sel_kwargs) -> JaxDataArray:
+    def sel(self, indexers: dict = None, method: str | None = None, **sel_kwargs) -> JaxDataArray:
         """Select a value from the :class:`.JaxDataArray` by indexing into coordinate values."""
+        if method is None:
+            return self._sel_exact(sel_kwargs)
+        elif method == "nearest":
+            return self._sel_nearest(sel_kwargs)
+        else:
+            raise NotImplementedError(f"Unkown selection method: {method}.")
+
+    def _sel_exact(self, sel_kwargs: dict) -> JaxDataArray:
         isel_kwargs = {}
-        for coord_name, sel_kwarg in sel_kwargs.items():
-            coord_list = self.get_coord_list(coord_name)
-            if isinstance(sel_kwarg, (tuple, list, np.ndarray)):
-                sel_kwarg = list(sel_kwarg)
-                isel_kwargs[coord_name] = []
-                for _sel_kwarg in sel_kwarg:
-                    if _sel_kwarg not in coord_list:
-                        raise DataError(
-                            f"Could not select '{coord_name}={_sel_kwarg}', value not found."
-                        )
-                    coord_index = coord_list.index(_sel_kwarg)
-                    isel_kwargs[coord_name].append(coord_index)
-            else:
-                if sel_kwarg not in coord_list:
-                    raise DataError(
-                        f"Could not select '{coord_name}={sel_kwarg}', value not found."
-                    )
-                coord_index = coord_list.index(sel_kwarg)
-                isel_kwargs[coord_name] = coord_index
+        for coord_name, vals in sel_kwargs.items():
+            coord_list = np.asarray(self.get_coord_list(coord_name))
+            vals = np.atleast_1d(vals)
+            dist = np.abs(coord_list[None] - vals[:, None])
+            indices = np.where(np.isclose(dist, 0))[1]
```

Comment:
Just so I understand, the only difference between _sel_exact and _sel_nearest seems to be how the indices are computed given the `coord_list` and `vals`

```py
# exact
vals = np.atleast_1d(vals)
dist = np.abs(coord_list[None] - vals[:, None])
indices = np.where(np.isclose(dist, 0))[1]
if indices.size == 0:
    raise DataError(f"Could not select '{coord_name}={vals}', some values were not found.")
```

```py
# nearest
vals = np.asarray(vals)
dist = np.abs(coord_list[:, None] - vals[None])
indices = np.argmin(dist, axis=0)
```

Is my understanding correct?

Furthermore, it could probably be written to just compute indices given some `dist` array.

Would it make sense then to refactor a bit to either just stick these in the same method (with an if / else) or just localize the `_get_indices(dist) -> incides` logic?

Reply:
Yes the calculation of the distance matrix should probably be refactored into its own method, you're right.

Reply:
well, after looking at it (see comments below) probably we can just keep the same distance method calculation, but we just need to refactor so that we compute the indices differently given the distance and some `method`?